File: 02_stack/03_generate_parenthesis/generate_parenthesis.py
from typing import List

class Solution:
    def generateParenthesis(self, n: int) -> List[str]:
        ans = list()
        solutions = list()

        def get_candidates(open: int, close: int) -> List[str]:
            candidates = []
            if (open < n):
                candidates.append("(")
            if (close < open):
                candidates.append(")")
            return candidates

        def backtrack(open: int, close: int):
            if open == close == n:
                solutions.append("".join(ans))
            for candidate in get_candidates(open, close):
                ans.append(candidate)
                backtrack(open + 1 if candidate == "(" else open,
                          close + 1 if candidate == ")" else close)
                ans.pop()

        backtrack(0, 0)
        return solutions


if __name__ == "__main__":
    solution = Solution()
    print(solution.generateParenthesis(3))

File: NeetCode/02_stack/03_generate_parenthesis/generate_parenthesis.cpp
// Standard: C++11

#include <vector>
#include <string>

using std::string;
using std::vector;

class Solution
{
  public:
    //------------------------------- Functions --------------------------------

    vector<string> generateParenthesis(int n)
    {
        vector<string> solutions;
        total_ = n;
        backtrack(0, 0, solutions);
        return solutions;
    }

  private:
    //------------------------------- Functions --------------------------------

    void backtrack(int open, int close, vector<string>& solutions)
    {
        if (open == close && open == total_)
            solutions.push_back(string(chars_.begin(), chars_.end()));
        vector<char> candidates = getCandidates(open, close);
        for (char candidate : candidates)
        {
            chars_.push_back(candidate);
            backtrack((candidate == '(') ? open + 1 : open,
                      (candidate == ')') ? close + 1 : close, solutions);
            chars_.pop_back();
        }
    }

    vector<char> getCandidates(int open, int close)
    {
        vector<char> candidates;
        if (open < total_)
            candidates.push_back('(');
        if (close < open)
            candidates.push_back(')');
        return candidates;
    }

    //---------------------------------- Data ----------------------------------

    vector<char> chars_;
    int          total_;
};

#include <iostream>

int main()
{
    Solution solution;
    auto     a = solution.generateParenthesis(3);
    for (auto each : a)
        std::cout << each << "\n";
}
